### PR TITLE
feat: support `bitwise` shift left/right

### LINKED ISSUE #2741: bitwise: support left/right shift using the native wrapping api
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The current of rust
https://doc.rust-lang.org/std/num/struct.Wrapping.html#

`wrapping_shl` and `wrapping_shr` is not stable.

the shift bit must be u32 data type

part of https://github.com/apache/arrow-rs/issues/2702

Now we just implement this use `value.wrapping_shl(u32)`, after the feature is stable, we can replace them.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Weijun-H**
# Which issue does this PR close?

Closes #2741

# Rationale for this change
Support `bitwise_shift_right` and `bitwise_shift_left`. And use 'as_usize' to ignore truncation.
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-arith/src/bitwise.rs [hunk 1]
```diff
@@ -143,6 +177,24 @@ mod tests {
         Ok(())
     }
 
+    #[test]
```
Comment (tustvold) [inline arrow-arith/src/bitwise.rs:180]: Is it worth adding a test of what happens if you shift by more than the number of bits?

**Review by tustvold — APPROVED** [review]

File: arrow-arith/src/bitwise.rs (hunk 1 above)
Comment (Weijun-H) [reply arrow-arith/src/bitwise.rs:180]: I add a test for it. It should be `0`

Comment (tustvold) [reply arrow-arith/src/bitwise.rs:180]: Are you sure? https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0bc4c3229b2ffd98a64eaf71573d0199

Comment (Weijun-H) [reply arrow-arith/src/bitwise.rs:180]: This makes sense, as `wrapping_shr` would shift the position would be the remainder of the type. So `wrapping_shr(3_i32, 32) == 3` and `wrapping_shr(3_i32, 33) == 1`

Comment (Weijun-H) [reply arrow-arith/src/bitwise.rs:180]: It seems like a rotation shift.

Comment (Weijun-H) [reply arrow-arith/src/bitwise.rs:180]: > Shift operations (<<, >>) on a value of width N can be passed a shift value >= N. It is unclear what behaviour should result from this, so the shift value is unconditionally masked to be modulo N to ensure that the argument is always in range.
I found the reason here https://github.com/rust-lang/rfcs/blob/master/text/0560-integer-overflow.md

File: arrow-arith/src/bitwise.rs [hunk 2]
```diff
@@ -17,7 +17,9 @@
 
 use crate::arity::{binary, unary};
 use arrow_array::*;
+use arrow_buffer::ArrowNativeType;
 use arrow_schema::ArrowError;
+use num::traits::{WrappingShl, WrappingShr};
```
Comment (alamb) [inline arrow-arith/src/bitwise.rs:22]: this is great -- when I looked into doing this in DataFusion I noticed these traits weren't in `std` -- 100% for finding them in `num`

**Review by alamb — COMMENTED** [review]
Thank you @Weijun-H  -- I can't wait to use this to simplify datafusio

**Pull request merged**